File: apps/app/src/model/kasho.ts
/** 不適切時に訳文上で示す誤りの範囲。文字は Unicode スカラー値の 0 始まり、半開区間 [start, end)。 */
export type Kasho = { shurui: "ayamari"; start: number; end: number };

/** 波線付き／なしの連続した断片。 */
export type KashoSegment = { text: string; ayamari: boolean };

function yakubunCharCount(yakubun: string): number {
  return [...yakubun].length;
}

/** 範囲外・逆転・ayamari 以外を捨てる。適切なら空にする。 */
export function normalizeKasho(
  yakubun: string,
  kasho: readonly Kasho[] | null | undefined,
  tekisetsu: boolean,
): Kasho[] {
  if (tekisetsu || kasho == null) {
    return [];
  }
  const len = yakubunCharCount(yakubun);
  const out: Kasho[] = [];
  for (const item of kasho) {
    if (item == null || typeof item !== "object") {
      continue;
    }
    if (item.shurui !== "ayamari") {
      continue;
    }
    if (
      Number.isInteger(item.start) &&
      Number.isInteger(item.end) &&
      item.start >= 0 &&
      item.start < item.end &&
      item.end <= len
    ) {
      out.push({ shurui: "ayamari", start: item.start, end: item.end });
    }
  }
  return out;
}

/** 訳文と箇所から、波線を付ける連続断片を作る。 */
export function buildKashoSegments(yakubun: string, kasho: readonly Kasho[]): KashoSegment[] {
  const chars = [...yakubun];
  if (chars.length === 0) {
    return [];
  }
  const cover = chars.map(() => false);
  for (const item of kasho) {
    for (let i = item.start; i < item.end; i++) {
      cover[i] = true;
    }
  }

  const segments: KashoSegment[] = [];
  for (let i = 0; i < chars.length;) {
    const ayamari = cover[i]!;
    let j = i + 1;
    while (j < chars.length && cover[j] === ayamari) {
      j += 1;
    }
    segments.push({ text: chars.slice(i, j).join(""), ayamari });
    i = j;
  }
  return segments;
}

File: apps/app/src-tauri/src/kasho.rs
use serde::de::Deserializer;
use serde::{Deserialize, Serialize};
use ts_rs::TS;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, TS)]
#[serde(tag = "shurui", rename_all = "camelCase")]
#[ts(export)]
pub enum Kasho {
    #[serde(rename = "ayamari")]
    Ayamari { start: usize, end: usize },
}

pub fn yakubun_char_count(yakubun: &str) -> usize {
    yakubun.chars().count()
}

pub fn normalize_kasho(yakubun: &str, kasho: Vec<Kasho>, tekisetsu: bool) -> Vec<Kasho> {
    if tekisetsu {
        return Vec::new();
    }
    let len = yakubun_char_count(yakubun);
    let mut out = Vec::new();
    for item in kasho {
        match item {
            Kasho::Ayamari { start, end } if start < end && end <= len => {
                out.push(Kasho::Ayamari { start, end });
            }
            _ => {}
        }
    }
    out
}

pub fn kasho_to_json(kasho: &[Kasho]) -> Result<String, String> {
    serde_json::to_string(kasho).map_err(|e| e.to_string())
}

/// 未知形（旧 ketsujo など）は無視して ayamari だけ残す。
pub fn kasho_from_json(raw: Option<String>) -> Result<Vec<Kasho>, String> {
    let Some(text) = raw else {
        return Ok(Vec::new());
    };
    if text.trim().is_empty() {
        return Ok(Vec::new());
    }
    let values: Vec<serde_json::Value> = serde_json::from_str(&text).map_err(|e| e.to_string())?;
    Ok(filter_ayamari(values))
}

pub fn deserialize_kasho<'de, D>(deserializer: D) -> Result<Vec<Kasho>, D::Error>
where
    D: Deserializer<'de>,
{
    let values = Vec::<serde_json::Value>::deserialize(deserializer)?;
    Ok(filter_ayamari(values))
}

fn filter_ayamari(values: Vec<serde_json::Value>) -> Vec<Kasho> {
    let mut out = Vec::new();
    for value in values {
        if let Ok(item) = serde_json::from_value::<Kasho>(value) {
            out.push(item);
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalize_keeps_valid_and_drops_invalid() {
        let out = normalize_kasho(
            "Hi",
            vec![
                Kasho::Ayamari { start: 0, end: 1 },
                Kasho::Ayamari { start: 1, end: 1 },
                Kasho::Ayamari { start: 0, end: 99 },
            ],
            false,
        );
        assert_eq!(out, vec![Kasho::Ayamari { start: 0, end: 1 }]);
    }

    #[test]
    fn json_roundtrip() {
        let items = vec![Kasho::Ayamari { start: 0, end: 2 }];
        let json = kasho_to_json(&items).unwrap();
        assert_eq!(kasho_from_json(Some(json)).unwrap(), items);
        assert!(kasho_from_json(None).unwrap().is_empty());
    }

    #[test]
    fn json_skips_unknown_shurui() {
        let raw = r#"[{"shurui":"ketsujo","index":1},{"shurui":"ayamari","start":0,"end":2}]"#;
        assert_eq!(
            kasho_from_json(Some(raw.to_string())).unwrap(),
            vec![Kasho::Ayamari { start: 0, end: 2 }]
        );
    }
}
